lib/db_utils.go

```diff
@@ -10661,6 +10681,409 @@ func DBDeletePostAssociationWithTxn(txn *badger.Txn, snap *Snapshot, association
+	var ret []*LockedBalanceEntry
+	var err error
+	handle.View(func(txn *badger.Txn) error {
+		ret, err = DBGetUnlockableLockedBalanceEntriesWithTxn(
+			txn, snap, hodlerPKID, profilePKID, currentTimestampUnixNanoSecs)
+		return nil
+	})
+	return ret, err
+}
+
+func DBGetUnlockableLockedBalanceEntriesWithTxn(
+	txn *badger.Txn,
+	snap *Snapshot,
+	hodlerPKID *PKID,
+	profilePKID *PKID,
+	currentTimestampUnixNanoSecs int64,
+) ([]*LockedBalanceEntry, error) {
+	// Retrieve all LockedBalanceEntries from db matching hodlerPKID, profilePKID, and
+	// ExpirationTimestampNanoSecs <= currentTimestampUnixNanoSecs.
+	// NOTE: While ideally we would start with <prefix, HODLerPKID, ProfilePKID> and
+	//       seek till <prefix, HODLerPKID, ProfilePKID, currentTimestampUnixNanoSecs>,
+	//       Badger does not support this functionality as the ValidForPrefix() function
+	//       stops when a mismatched prefix occurs, not a "lexicographically less than" prefix.
+	//       For this reason, we start with <prefix, HODLerPKID, ProfilePKID, currentTimestampUnixNanoSecs>
+	//       and iterate backwards while we're valid for the prefix <prefix, HODLerPKID, ProfilePKID>.
+
+	if currentTimestampUnixNanoSecs < 0 || currentTimestampUnixNanoSecs == math.MaxInt64-1 {
+		return nil, fmt.Errorf("DBGetUnlockableLockedBalanceEntriesWithTxn: invalid " +
+			"block timestamp; this shouldn't be possible")
+	}
+
+	// Start at <prefix, HODLerPKID, ProfilePKID, CurrentTimestampNanoSecs>
+	startKey := _dbKeyForLockedBalanceEntry(LockedBalanceEntry{
+		HODLerPKID:                  hodlerPKID,
+		ProfilePKID:                 profilePKID,
+		ExpirationTimestampNanoSecs: currentTimestampUnixNanoSecs,
+	})
+
+	// Valid for prefix <prefix, HODLerPKID, ProfilePKID>
+	prefixKey := DBPrefixKeyForLockedBalanceEntryByHODLerPKIDandProfilePKID(&LockedBalanceEntry{
+		HODLerPKID:  hodlerPKID,
+		ProfilePKID: profilePKID,
+	})
+
+	// Create an iterator. We set the iterator to reverse in o
+	opts := badger.DefaultIteratorOptions
+	opts.Reverse = true
+	iterator := txn.NewIterator(opts)
+	defer iterator.Close()
+
+	// Store matching LockedBalanceEntries to return
+	var lockedBalanceEntries []*LockedBalanceEntry
+
+	// Loop.
+	for iterator.Seek(startKey); iterator.ValidForPrefix(prefixKey); iterator.Next() {
+		// Retrieve the LockedBalanceEntryBytes.
+		lockedBalanceEntryBytes, err := iterator.Item().ValueCopy(nil)
+		if err != nil {
+			return nil, errors.Wrapf(err, "DBGetUnlockableLockedBalanceEntriesWithTxn: "+
+				"error retrieveing LockedBalanceEntry: ")
+		}
+
+		// Convert LockedBalanceEntryBytes to LockedBalanceEntry.
+		rr := bytes.NewReader(lockedBalanceEntryBytes)
+		lockedBalanceEntry, err := DecodeDeSoEncoder(&LockedBalanceEntry{}, rr)
+		if err != nil {
+			return nil, errors.Wrapf(err, "DBGetUnlockableLockedBalanceEntriesWithTxn: "+
+				"error decoding LockedBalanceEntry: ")
+		}
+
+		if lockedBalanceEntry.ExpirationTimestampNanoSecs < currentTimestampUnixNanoSecs {
```

Comment:
* Shouldn't it always be the case that ExpirationTimestampNanoSecs is less than currentTimestampUnixNanosSecs, since that's how you constructed the seek? If yes, add a comment explaining that this is a sanity check. If not, explain in a comment why this if check is needed.
* Separate question: It looks like this will ignore things when they're EQUAL to currentTimestampUnixNanoSecs. It's an edge-case, but just wanted to highlight it. Maybe mention in a comment that this is intended.